### DISCUSSION TITLE
Added support for Ethereum Classic (ETC) & UBIQ (UBQ)

### DISCUSSION THREAD

**Discussion by blackmennewstyle** [coinfoundry/miningcore#1586] [General]
Hi everyone,

Following my initial pull request https://github.com/oliverw/miningcore/pull/1514 which added support for Conceal https://github.com/oliverw/miningcore/discussions/1507#discussioncomment-4222587, i have added support for `Ethereum Classic (ETC)` and also `Ubiq (UBQ)` .

Following the previous amazing work from @jon4hz regarding `Uncle` and `Orphaned` https://github.com/oliverw/miningcore/pull/1501, i also fixed an annoying bug in the `EthereumPayoutHandler.cs`, which technically impacts all `Ethereum (Uncle) blockchains`.
In a very particular case-scenario, when a miner/pool has enough `hash-rate-power` and is able to mine two blocks with the same `blockHeight` https://github.com/oliverw/miningcore/issues/1583#issuecomment-1383009149.
They are always respectively an `Uncle` block and `Regular` block.
I have seen some people deleting one of them in their bug report, which is a huuuge mistake, since they have both a `block reward` and deleting one is basically: stealing coins from miner(s) :'(
The issue has been reported multiple times https://github.com/oliverw/miningcore/issues/1080#issue-1098003849 but it's very difficult to reproduce on most `mainnet` but it is very easy to encounter on some `testnet` like on ETC for example.

**How to setup `miningcore` for ETC Mordor `testnet`**

`geth node`
```
geth --mordor --config /home/ceedii/.ethereum/mordor.toml --http --mine --miner.threads 1 --unlock 0x421Afb2ce225D3A2d3DD6e63Fe57E124B40e20Af --password /home/ceedii/.ethereum/mordor/help/pool.password --allow-insecure-unlock --verbosity 4 --log.debug > /home/ceedii/.ethereum/mordor.log 2>&1
```
(Please refer to the online documentation - https://geth.ethereum.org/ - for further assistance if any doubt)

By default the `etchash` implementation i made in `miningcore`, based on this user contribution https://github.com/oliverw/miningcore/discussions/1415#discussioncomment-4561890, comes ready for `mainnet`.
For being able to mine on the `ETC` Mordor `testnet`, it is **mandatory** to change the 2 following two lines accordingly, in the file `miningcore/src/Native/libetchash/ethash.h`:
```
//#define ETCHASH_FORK_BLOCK 11700000 // classic mainnet
#define ETCHASH_FORK_BLOCK 2520000 // mordor
```
You will need then to compile or recompile `miningcore`.

`miningcore config.json`
An `ethereumclassic_pool.json` file is provided in the folder `miningcore/examples/`. However, it is **mandatory** to use:
```
"chainTypeOverride": "Mordor",
```
Lastly, **do not forget** to fill that section:
```
"dagDir": "<Your directory for the dag file>",
```
The following applies only if you are running your `node` on the same `server/machine` with `miningcore` and want to take advantage of `geth` automatically generating the `DAG` files.: If you use a `.toml` file just like i did, the answer is inside this file.

**In any other cases, or if you are not sure, just fill that option with** `""` => `"dagDir": "",`

**How to setup `miningcore` for ETC  `mainnet`**

`geth node`
```
geth --classic --config /home/ceedii/.ethereum/classic.toml --http --mine --miner.threads 1 --unlock 0x421Afb2ce225D3A2d3DD6e63Fe57E124B40e20Af --password /home/ceedii/.ethereum/classic/help/pool.password --allow-insecure-unlock --verbosity 4 --log.debug > /home/ceedii/.ethereum/classic.log 2>&1
```
(Please refer to the online documentation - https://geth.ethereum.org/ - for further assistance if any doubt)

By default the `etchash` implementation i made in `miningcore`, based on this user contribution https://github.com/oliverw/miningcore/discussions/1415#discussioncomment-4561890, comes ready for `mainnet`.
But in case you previously made the change for the `ETC` Mordor `testnet`, it is **mandatory** to change the 2 following two lines accordingly, in the file `miningcore/src/Native/libetchash/ethash.h`:
```
#define ETCHASH_FORK_BLOCK 11700000 // classic mainnet
//#define ETCHASH_FORK_BLOCK 2520000 // mordor
```
You will need then to compile or recompile `miningcore`.

`miningcore config.json`
An `ethereumclassic_pool.json` file is provided in the folder `miningcore/examples/`. By default it comes already setup for `mainnet`:
```
"chainTypeOverride": "Classic",
```
Lastly, **do not forget** to fill that section:
```
"dagDir": "<Your directory for the dag file>",
```
The following applies only if you are running your `node` on the same `server/machine` with `miningcore` and want to take advantage of `geth` automatically generating the `DAG` files: If you use a `.toml` file just like i did, the answer is inside this file.

**In any other cases, or if you are not sure, just fill that option with** `""` => `"dagDir": "",`

**How to setup `miningcore` for UBIQ  `mainnet`**

`gubiq node` (gubiq is a fork of `geth`)
```
gubiq --config /home/ceedii/.ubiq/mainnet.toml --http --mine --miner.threads 1 --unlock 0x6101Ee22AB2cd3907E5CA0b70EF12A2afE842A56 --password /home/ceedii/.ubiq/mainnet/help/pool.password --allow-insecure-unlock --verbosity 4 --log.debug > /home/ceedii/.ubiq/mainnet.log 2>&1
```
(Since `gubiq` is a fork of `geth`, please refer to the online documentation - https://geth.ethereum.org/ - for further assistance if any doubt)

`miningcore config.json`
An `ubiq_pool.json` file is provided in the folder `miningcore/examples/`. By default it comes already setup for `mainnet`:
```
"chainTypeOverride": "Ubiq",
```
Lastly, **do not forget** to fill that section:
```
"dagDir": "<Your directory for the dag file>",
```
The following applies only if you are running your `node` on the same `server/machine` with `miningcore` and want to take advantage of `gubiq` automatically generating the `DAG` files: If you use a `.toml` file just like i did, the answer is inside this file.

**In any other cases, or if you are not sure, just fill that option with** `""` => `"dagDir": "",`

**How to setup `miningcore` for UBIQ  `testnet`**
Sadly, there is no POW Ubiq `testnet` or should i say there is one but it is not open to the public. Apparently, it suffered from previous users abuses and it will not be open again to the public anytime soon.
Well, at least that was the experience and feedback i received when i tried to ask for access on `reddit` and also `discord`.

So hope you enjoy my modest contribution and happy `block` hunting :fire: :rocket: 

![Screenshot_2023-01-21_07-59-05](https://user-images.githubusercontent.com/4682874/213855750-b5c1b5af-c81a-46be-95e2-13fa72f466e0.png)
(My modest UBIQ (UBQ) mining pool: https://ubiq.cedric-crispin.com/)

**Comment by Censseo** [comment]
Amazing work. I saw your PR few days ago and tested it on Mordor, it works, but I was facing kind of same issues than with open etc pool. I don't feel good using geth in an unsafe way, and with docker it doesn't want to unlock at all, still didn't figure out why. And second thing, I don't get why there is a dagfile, it completely crush the thing at it takes ages to generate, more than 30 min at each start, I still don't understand if there is an issue on this on my side and I think it is not relative to your PR.
Anyway good job and thks for the work 🙂

**Reply by blackmennewstyle** [reply]
Thank you. I never tried to run `miningcore` through docker but i think the reason we are using `unsafe` on `geth` is maybe because of the same reason we are disabling `rpc-login` with `monero`, the `C# library` used in `miningcore` can't do identification in some context if i am not mistaken.

It needs the dag files for sending jobs and accepting the shares it receives, if you are running `miningcore` and `geth` on the same server, you can actually directly put the folder path where `geth` generates them in the `config.json`: `"dagDir": "/home/ceedii/.etchash/",` that will skip the generation on the `miningcore` side since `geth` automatically generates them.

You might want maybe to update your `miningcore` repo since i made some modifications recently.

**Reply by Censseo** [reply]
I run all of my mining pool services on docker with docker swarm. I would be happy to share with you about that, you should come to our discord: https://discord.gg/B9uxQHgp :)
and our pool: https://jjpool.fr

**Reply by jon4hz** [reply]
Actually the DAG isn't 100% necessary. For example open-ethereum-pool does validation with a light cache only. 
I made a first test to do the same with miningcore here: https://github.com/jon4hz/miningcore/commit/ac71e5755aedb591122e454c67e1f7c02fa73526

However I'm not sure if there are any downsides to using a light cache only. It would make operating a pool much easier but I don't know if there is a performance impact.

**Reply by blackmennewstyle** [reply]
Oh yes my bad, for some reason i convinced myself it is, i will modify my documentation, thanks for the heads-up as always :)
